feat(mcdu): upgrade perf appr page to release 2

tracernz opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
~~Partial #1823~~ ~~(still need go-around page, to come in future PR, with changes from #1824)~~ Done in #3650.
Fixes #2096
Fixes #3121

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
- Pull approach page changes from @beheh's previous PR (#1824)
- Tighten/fix input checks per neo FCOM entry format listing
- Remember pilot entered values and show in large font (automatic/calculated values small)
- Rework approach page appearance to Honeywell release 2 standard
- ~~Allow selection of BARO-RADIO or MDA-DH for minimums~~

## Questions
- Can QNH, Temp, Wind be cleared once entered?
- ~~Does LDG CONF look correct when CONF3 selected? (no references)~~
- ~~Does FLS/FINAL APP look correct when FINAL APP selected? (note FLS is inop at this stage, so greyed out)~~
- ~~What happens when QNH is already entered, and hPa/inHg selection is changed on EFIS?~~

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
Empty page, no approach selected:
![appr_empty](https://user-images.githubusercontent.com/9995998/110750589-966ff100-82a7-11eb-89a4-35b4cb62a673.PNG)
Empty page with approach selected:
![appr_empty_radio](https://user-images.githubusercontent.com/9995998/110750650-ac7db180-82a7-11eb-81a9-7cf0988226ec.PNG)
Empty page within 180 nm of destination:
![appr_empty_within180](https://user-images.githubusercontent.com/9995998/110750666-b0a9cf00-82a7-11eb-901b-83cdfe196dc9.PNG)
Filled page with CONF3 selected:
![appr_filled_conf3](https://user-images.githubusercontent.com/9995998/110750682-b6071980-82a7-11eb-8127-ed7d3961e7d3.PNG)
Filled page with non-precision approach selected:
![appr_filled_nonprec](https://user-images.githubusercontent.com/9995998/110750692-bacbcd80-82a7-11eb-926b-738d21eb417d.PNG)
Pilot entered Vapp:
![appr_pilot_vapp](https://user-images.githubusercontent.com/9995998/110750706-bf908180-82a7-11eb-91e5-de902a93c357.PNG)

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->
[A330 with release 2](https://youtu.be/MuP0_u6Q778?t=62):
![image](https://user-images.githubusercontent.com/9995998/108594984-1cec8d80-73e2-11eb-9209-fb1f80108343.png)
FCOM (release 2):
![image](https://user-images.githubusercontent.com/9995998/108595026-5f15cf00-73e2-11eb-99c8-9ccfb3d3ff9e.png)
FCOM (step 1a):
![image](https://user-images.githubusercontent.com/9995998/108595307-1101cb00-73e4-11eb-94d4-fa121908674e.png)
![image](https://user-images.githubusercontent.com/9995998/108595047-7d7bca80-73e2-11eb-941f-114b4e90c6f3.png)
![image](https://user-images.githubusercontent.com/9995998/108595068-a603c480-73e2-11eb-8b5d-2734915dda6d.png)


## Additional context
<!-- Add any other context about the pull request here. -->
~~Note an extra column between the left and centre columns used for VLS in this new format. Please consider this for your react MCDU work @Lucky38i.~~ Not required with new font.

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): tracernz

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
- Spawn a new flight at the gate and go to PERF APPR page
  - Try various entry formats e.g. 29.92 (in.Hg), 2992 (in.Hg), 1013 (hPa), and variations thereof, including 3 digit hPa, also illegal ones you can think of, and make sure you can't accidentally enter a temperature
  - Switch LDG FLAPs config and observe the lines VLS and VAPP updating appropriately
  - Override VAPP and ensure the font goes large for pilot entry, then clear to revert to automatic calculation
  - Repeat with TRANS ALT
  - Enter various winds, valid and invalid to ensure correct display, with leading zeroes.
- Spawn a flight a little over 180 Nm from destination, enter flightplan but don't enter APPR values.
  - Check that QNH and TEMP turn to amber boxes when 180 Nm out
  - Select ILS approach on F-PLN and ensure both BARO and RADIO entries are available on APPR, FINAL APP/FLS are hidden, and page title shows selected approach appropriately
  - Select non-precision approaches (RNAV, VOR-DME, etc.), check that only BARO entry is available, and page title shows selected approach appropriately
- Fly an approach or two and make sure you're happy!

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js:
```diff
@@ -774,47 +790,68 @@ class CDUPerformancePage {
         }
-        let mdaCell = "[ ]";
+
+        const baroRadio = NXDataStore.get("CONFIG_BARO_RADIO", "0") === "1";
```
I'd make a variable for this and set it to the selected state on init and when changed on the options page.

tracernz commented:
Notes to self:
- should be able to align VLS using centre column once we have a proper fixed-width font from #3650 
- Change log

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js:
```diff
@@ -706,34 +707,45 @@ class CDUPerformancePage {
-        let magWindSpeedCell = "[ ]";
+        let magWindSpeedCell = "[\xa0]";
         if (isFinite(mcdu.perfApprWindSpeed)) {
             magWindSpeedCell = mcdu.perfApprWindSpeed.toFixed(0);
```
Looks like this should also be padLeft'ed, so that it's 001 (according to your small reference screenshot).

tracernz commented:
I have rebased this after the new font was merged, and was able to simplify things a bit as a result of the fixed width. Also did both feedback items, many thanks 👍 .

Bonus: how it looked before and after @tyler58546's font change :rocket: (and a couple of my tweaks):
![appr_filled_conf3](https://user-images.githubusercontent.com/9995998/108594907-95068380-73e1-11eb-87b4-2b44420b65b1.PNG)
![appr_filled_conf3](https://user-images.githubusercontent.com/9995998/108972046-ccc64180-76e7-11eb-8863-8885685911ae.PNG)

beheh commented:
Looks great so far!

I'm a little on the fence whether we want to offer the MDA/BARO setting at all. We've had long discussions about this on Discord, especially around how that setting seems to sometimes be related to the existence of the QNH/QFE switch we have, but the general consensus seems to be that Neos are nearly always on the BARO/RADIO. I think it might be easier to go with BARO/RADIO without an option and call it a day.

tracernz commented:
> I'm a little on the fence whether we want to offer the MDA/BARO setting at all. We've had long discussions about this on Discord, especially around how that setting seems to sometimes be related to the existence of the QNH/QFE switch we have, but the general consensus seems to be that Neos are nearly always on the BARO/RADIO. I think it might be easier to go with BARO/RADIO without an option and call it a day.

The reason I added that option is because Tarek is very keen on it. I am happy with BARO/RADIO but if people really want MDA/DH I think it's worth an option.

Here's the FCOM description (for step 1a neo...):
![image](https://user-images.githubusercontent.com/9995998/109448548-14afe480-7aab-11eb-8473-49c8e4126eea.png)

tracernz commented:
> 
> 
> Looks great so far!

^_^
> 
> I'm a little on the fence whether we want to offer the MDA/BARO setting at all. We've had long discussions about this on Discord, especially around how that setting seems to sometimes be related to the existence of the QNH/QFE switch we have, but the general consensus seems to be that Neos are nearly always on the BARO/RADIO. I think it might be easier to go with BARO/RADIO without an option and call it a day.

Removed the option. It can always be added later if deemed necessary.

tracernz commented:
Rebased, updated screenshots and testing instructions, added a new commit to explicitly prevent CLR key QNH/TEMP/WIND (before it gave a format error). Should be ready for review I think.

MisterChocker reviewed (approved):
Code wise LGTM

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 13.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3643
Tier of testing: 1
Changes to observe: APPR PERF page
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Tested entry formats and limits

- QNH 22.00-31.00 (Dail goes from 27.99-32.01), 750-1050
- TRANS ALT goes to 99999
- VAPP 90-350
- BARO has no limits (negative or positive) (Image1)

- Runway info works
- CONF3/FULL works
- Manual VAPP works

Media:

![134](https://user-images.githubusercontent.com/78707844/111014202-e7b0db00-83a2-11eb-8e93-6262470ce0d0.png)
Image1

Issues: Small issue with BARO limits 

Overall Rating: -

Conclusions of the testing: Did just the part1 test so far. Maybe needs a recheck for some limits.

Hope this helps :-)

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js:
```diff
@@ -784,47 +796,64 @@ class CDUPerformancePage {
+            baroCell = mcdu.perfApprMDA.toFixed(0);
         }
         mcdu.onRightInput[1] = (value) => {
             if (mcdu.setPerfApprMDA(value) && mcdu.setPerfApprDH(FMCMainDisplay.clrValue)) {
```
As mentioned already, this needs to be limited to (airfield elevation to airfield elevation + 5000ft) and (0 to 5000) respectively.

MisterChocker replied:
iircf clamp(0, 5000, value)

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2476,59 +2477,90 @@ class FMCMainDisplay extends BaseAirliners {
-            heading = parseFloat(sSplit[0]);
-            speed = parseFloat(sSplit[1]);
+        if (s === FMCMainDisplay.clrValue) {
+            this.addNewMessage(NXSystemMessages.notAllowed);
```
I would expect this field to be clearable at all times. We may need pilot confirmation?

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2476,59 +2477,90 @@ class FMCMainDisplay extends BaseAirliners {
 
     setPerfApprQNH(s) {
+        if (s === FMCMainDisplay.clrValue) {
+            this.addNewMessage(NXSystemMessages.notAllowed);
```
My theory is that this can be cleared as long as you're the 180 NM out and the PERF APPR page shows cyan brackets. Once you're closer though, and you get the amber boxes, I believe it should be able to clear it anymore. We should probably get pilot confirmation on this and have someone test it.

tracernz replied:
Yes, I did ask for that but not heard anything back yet. I’ll try again. We didn’t support it before so I defaulted to not allowing until I heard something different. I think you are right though, following the general rules, only amber boxes are normally not clearable.

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -2476,59 +2477,90 @@ class FMCMainDisplay extends BaseAirliners {
-            this.perfApprTemp = value;
-            return true;
+        if (s === FMCMainDisplay.clrValue) {
+            this.addNewMessage(NXSystemMessages.notAllowed);
```
I would expect this to follow the logic of the QNH field, which could mean "clearable unless you're within 180 NM and have amber boxes on the PERF APPR page".

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js:
```diff
@@ -754,9 +762,13 @@ class CDUPerformancePage {
+            }
         }
         mcdu.onLeftInput[3] = (value) => {
             if (mcdu.setPerfApprTransAlt(value)) {
```
This needs to follow the logic from the TAKEOFF page: 0-45000ft, rounded to nearest 10ft.

tracernz replied:
Takeoff page is 1000 - 45000, nearest 10. I've taken that logic (will push maybe this evening when I get a chance to work on the other bits).

beheh commented:
Added a few comments, mostly things that need accuracy checks. Really looking forward to seeing this in the sim!

tracernz commented:
Rebased and addressed testing + review feedback (last commit). Thanks to you all. 👍 

For clearing of QNH/temp/wind, I have allowed clearing of QNH and temp (the two that turn amber) if further than 180 to destination, and wind can be cleared at any time. I have asked for more info in #pilot-feedback so I may need to change again if anything different comes back. I have also disallowed calls to set DH when a non-ILS approach is selected, per the FCOM, but I think this is a moot point anyway, because the entry is not shown.

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 18.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3643
Tier of testing: 1
Changes to observe: PERF APPR page
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Set up simple F-PLN to EDDM
3.) Test Perf APPR page
- Baro got limitations
- Entries were clearable

4.) Took off to EDDM
- at 180NM QNH and TEMP become boxed and are not clearable

Issues: -

Overall Rating: Good

Conclusions of the testing: Seems to work

Hope this helps :-)

FlightSimmer7700 commented:
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 20/3/21
Version of the sim: 1.14.5.0
PR Tested: #3643 
Tier of testing: 1
Changes to observe:
approach page

Testing technique: Did a full flight
Steps you've made to test the Pull Request:
1. Loaded in and did everything listed
2. reloaded in and did a full flight
3. forgot to test everything properly while doing the full flight (flying into LOWI)
4. will need to do a further test on another full flight

Media: Insert screenshots, recordings, or anything relevant to the review here.
![image](https://user-images.githubusercontent.com/70103461/111854837-48976080-8919-11eb-8e70-b3cdda199b41.png)
![image](https://user-images.githubusercontent.com/70103461/111854840-4c2ae780-8919-11eb-9b93-4b10ede45773.png)
![image](https://user-images.githubusercontent.com/70103461/111854845-4fbe6e80-8919-11eb-9470-f6ed09fed8a4.png)

Issues (If Any): 

- QNH field is "broken" and can go to very low values but not as high and I don't think they need to be going that low anyway unless that is how it is irl

Overall Rating: To be improved
Conclusions of the testing: approach page now looks much better with only qnh to be fixed for now

tracernz commented:
>QNH field is "broken" and can go to very low values but not as high and I don't think they need to be going that low anyway unless that is how it is irl

QNH limits are based on this neo FCOM listing:
![image](https://user-images.githubusercontent.com/9995998/111864921-b5b5ff00-89c8-11eb-862d-291aa4ec8c3e.png)

FlightSimmer7700 commented:
> > QNH field is "broken" and can go to very low values but not as high and I don't think they need to be going that low anyway unless that is how it is irl
> 
> QNH limits are based on this neo FMS2 release 1a FCOM listing (don't have any release 2 references for this). The option is QFE, which we simulate:
> ![image](https://user-images.githubusercontent.com/9995998/111864921-b5b5ff00-89c8-11eb-862d-291aa4ec8c3e.png)

oh ok sorry

tracernz commented:
> 
> 
> > > QNH field is "broken" and can go to very low values but not as high and I don't think they need to be going that low anyway unless that is how it is irl
> > 
> > 
> > QNH limits are based on this neo FMS2 release 1a FCOM listing (don't have any release 2 references for this). The option is QFE, which we simulate:
> > ![image](https://user-images.githubusercontent.com/9995998/111864921-b5b5ff00-89c8-11eb-862d-291aa4ec8c3e.png)
> 
> oh ok sorry

No problem. If you ever think something is not right, please speak up! Either we will find a problem, or satisfy ourselves it is correct. Either way, it's a win. 😁
